zerver/tests/test_auth_backends.py

```diff
@@ -942,6 +945,151 @@ def test_social_auth_complete_when_base_exc_is_raised(self) -> None:
+
+    BACKEND_CLASS = SAMLAuthBackend
+    CLIENT_KEY_SETTING = "SOCIAL_AUTH_SAML_SP_PRIVATE_KEY"
+    LOGIN_URL = "/accounts/login/social/saml"
+    SIGNUP_URL = "/accounts/register/social/saml"
+    AUTHORIZATION_URL = "https://idp.testshib.org/idp/profile/SAML2/Redirect/SSO"
+    AUTH_FINISH_URL = "/complete/saml/"
+    CONFIG_ERROR_URL = "/config-error/saml"
+
+    # We have to define our own social_auth_test as the flow of SAML authentication
+    # is different from the other social backends.
+    def social_auth_test(self, account_data_dict: Dict[str, str],
+                         *, subdomain: Optional[str]=None,
+                         mobile_flow_otp: Optional[str]=None,
+                         is_signup: Optional[str]=None,
+                         next: str='',
+                         multiuse_object_key: str='',
+                         **extra_data: Any) -> HttpResponse:
+        url = self.LOGIN_URL
+
+        params = {}
+        headers = {}
+        if subdomain is not None:
+            headers['HTTP_HOST'] = subdomain + ".testserver"
+        if mobile_flow_otp is not None:
+            params['mobile_flow_otp'] = mobile_flow_otp
+            headers['HTTP_USER_AGENT'] = "ZulipAndroid"
+        if is_signup is not None:
+            url = self.SIGNUP_URL
+        params['next'] = next
+        params['multiuse_object_key'] = multiuse_object_key
+        if len(params) > 0:
+            url += "?%s" % (urllib.parse.urlencode(params),)
+
+        result = self.client_get(url, **headers)
+
+        expected_result_url_prefix = 'http://testserver/login/%s/' % (self.backend.name,)
+        if settings.SOCIAL_AUTH_SUBDOMAIN is not None:
+            expected_result_url_prefix = (
+                'http://%s.testserver/login/%s/' % (settings.SOCIAL_AUTH_SUBDOMAIN, self.backend.name)
+            )
+
+        if result.status_code != 302 or not result.url.startswith(expected_result_url_prefix):
+            return result
+
+        result = self.client_get(result.url, **headers)
+
+        self.assertEqual(result.status_code, 302)
+        assert self.AUTHORIZATION_URL in result.url
+        assert "samlrequest" in result.url.lower()
+
+        self.client.cookies = result.cookies
+
+        saml_response = self.generate_saml_response(**account_data_dict)
+        post_params = {"SAMLResponse": saml_response, "RelayState": "test_idp"}
+        # The mock below is necessary, so that python3-saml accepts our SAMLResponse,
+        # and doesn't verify the cryptographic signatures etc., since generating
+        # a perfectly valid SAMLResponse for the purpose of these tests would be too complex,
+        # and we simply use one loaded from a fixture file.
+        with mock.patch.object(OneLogin_Saml2_Response, 'is_valid', return_value=True):
+            result = self.client_post(self.AUTH_FINISH_URL, post_params, **headers)
+
+        return result
+
+    def generate_saml_response(self, email: str, name: str) -> str:
```

Comment:
It's probably worth writing a comment string explaining this function's role.

Reply:
Done:
```
    def generate_saml_response(self, email: str, name: str) -> str:
        """
        The samlresponse.txt fixture has a pre-generated SAMLResponse,
        with {email}, {first_name}, {last_name} placeholders, that can
        be filled out with the data we want.
        """
```